### ISSUE TITLE
[1.9.x] Sometimes multipart body missed last 1 byte.

### ISSUE THREAD

**Issue by carryel** [AsyncHttpClient/async-http-client#1271]
MultipartBody#read reads multipart body and writes packets in a given ByteBuffer. 

For examples, 

If ByteBuffer's size is 8192 and remaining multipart body's size is 8193, last 1 byte(\n) will be missed.

In MultipartBody.java,

```java
    private boolean currentBytesFullyRead() {
        return currentBytes == null || currentBytesPosition >= currentBytes.length - 1;
    }

```

Maybe currentBytes should be written in ByteBuffer and currentBytesPosition means the positon of current bytes after writing.

And it seems that currentBytesFullyRead() means whether currentBytes are written in ByteBuffer totally or not.

Actually "currentBytesPosition == currentBytes.length - 1" means the last byte of currentBytes should be written in next ByteBuffer later.

But currentBytesFullyRead returns true, so currentBytes, currentBytesPosition and transferDone will be reset(missing one byte, '\n').

Please see the next source code:

In MultipartBody.java,
```java
public long read(ByteBuffer buffer) throws IOException {
    ...
            if (doneWritingParts) {
                if (currentBytesPosition == -1) {
                    initializeCurrentBytes(MultipartUtils.getMessageEnd(boundary));
                }

                if (currentBytesPosition > -1) {
                    overallLength += writeCurrentBytes(buffer, maxLength - overallLength);

                    if (currentBytesFullyRead()) {
                        // reset and missed 1 bytes.
                        currentBytes = null;
                        currentBytesPosition = -1;
                        transfertDone = true;
                    }
                }
            }
    ...
```

So I think MultipartBody#currentBytesFullyRead will be fixed like below:

```java
    private boolean currentBytesFullyRead() {
        return currentBytes == null || currentBytesPosition == -1;
    }
```

Sometimes I met one missing packet at the upload server when I used AHC with multipart.

When I test it again with the proposed patch, I can find the problem is resolved.

**Comment by slandelle** [comment]
Can you reproduce with AHC2?

**Comment by carryel** [comment]
@slandelle O.K. I will try this issue with AHC2. I think AHC2 will be fine.

**Comment by carryel** [comment]
@slandelle When I tested with AHC2, I couldn't reproduce this issue because AHC2 sent body to WritableByteChannel directly via ByteBuffer.

Unfortunately, I can't migrate AHC1.9.x into AHC2 in less time because there are a lot of works and I would like to keep the grizzly provider.

So, could you fix and release 1.9.x because my production environment met this problem occasionally? 

The writeCurrentBytes() already considers resetting position and bytes below:

```java
    private int writeCurrentBytes(ByteBuffer buffer, int length) throws IOException {
        ...
        if (writeLength > 0) {
            buffer.put(currentBytes, currentBytesPosition, writeLength);

            if (available <= length) {
                // reset position and bytes
                currentBytesPosition = -1;
                currentBytes = null;
            } else {
                currentBytesPosition += writeLength;
            }
        }

        return writeLength;
    }
```

So I think the proposed patch is not side-effect. (Of course I tested normal cases)

**Comment by slandelle** [comment]
Could you please provide a PR and a test demonstrating the issue?
AHC1 has reached EOL, so you'd better consider upgrading.

**Comment by carryel** [comment]
@slandelle Sure, I will.

**Comment by carryel** [comment]
@slandelle I also added the testcase for reproducing this issue.
Before the patch, you can confirm the following error:
```
Failed tests:   transferWithCopy2(com.ning.http.client.multipart.MultipartBodyTest): expected [650] but found [649]
```

**Comment by slandelle** [comment]
Fixed by #1275
